_posts/2020-03-30-method-s-arguments-pt-2.markdown

```diff
@@ -0,0 +1,350 @@
+---
+layout: post
+title: "Exploring Method Arguments in Ruby: Part 2"
+date: 2020-04-05 09:30:00
+categories: [ruby, learning]
+author: arieljuod
+---
+
+On [the first part of this series](https://www.ombulabs.com/blog/ruby/learning/method-s-arguments-pt-1.html "Exploring Method Arguments in Ruby: Part 1 - The Lean Software Boutique") we talked about positional arguments, but there are more types. In this second part we'll talk about keyword arguments.
+
+Positional and keyword arguments share a lot of characteristics so we'll split this article in similar sections, but there are some key differences we'll point on each section.
+
+<!--more-->
+
+## Keyword arguments
+
+The main difference between positional and keyword arguments is that the former have to be passed in a specific order while keyword arguments can be used in any order using a different syntax.
+
+## Required keyword arguments
+
+These keyword arguments are required when calling a method. If the method defines required parameters you have to provide an argument for each key. The syntax when defining required keyword arguments for a method is similar to positional arguments except you need a colon at the end of the parameter name. Like this:
+
+```ruby
+def foo(arg1:)
+  puts "arg1 is: #{arg1.inspect}"
+end
+```
+
+And you have to add key value pairs matching the keys defined on the method definition when calling the method. Keys MUST be symbols.
+
+```ruby
+foo(arg1: 1) # you have to specify the key / parameter name
+# => arg1 is: 1
+
+foo(:arg1 => 1) # you can use the old hash syntax
+# => arg1 is: 1
+
+foo
+# ArgumentError (missing keyword: arg1)
+
+foo(1) # notice ruby knows that it expects 0 positinal arguments but requires keyword arguments
+# ArgumentError (wrong number of arguments (given 1, expected 0; required keyword: arg1))
+
+```
+
+You can assign the same key multiple times and the last one will be the value for that parameter. You'll see a warning though (but no error):
+
+```ruby
+foo(arg1: 1, arg1: 2)
+# => warning: key :arg1 is duplicated and overwritten on line X
+#    arg1 is: 2
+```
+
+Up until ruby 2.6 you can wrap keyword arguments with curly braces with no issues. Ruby will deconstruct the hash into keyword arguments transparently:
+
+```ruby
+# ruby 2.6.3
+foo({arg1: 1})
+# => arg1 is: 1
+```
+
+But this WON'T be allowed on ruby 3 and will raise an **ArgumentError** exception. If you use this with ruby 2.7 you'll see a deprecation warning:
```

Comment:
@arielj Maybe add a link to this page? https://www.ruby-lang.org/en/news/2019/12/12/separation-of-positional-and-keyword-arguments-in-ruby-3-0/